src/qtism/runtime/tests/RouteItem.php

```diff
@@ -450,4 +450,68 @@ public function getTimeLimits($excludeItem = false): RouteTimeLimitsCollection
 
         return $timeLimits;
     }
+
+    public function getEffectiveBranchRules(): BranchRuleCollection
+    {
+        if ($this->getBranchRules()->count() > 0) {
+            return $this->getBranchRules();
+        }
+
+        $sectionBranchRules = $this->getEffectiveSectionBranchRules();
+
+        if ($sectionBranchRules === null || $sectionBranchRules->count() > 0) {
+            return $sectionBranchRules ?? new BranchRuleCollection();
+        }
+
+        $testPartSections = $this->getTestPart()->getAssessmentSections()->getArrayCopy();
+
+        if (
+            end($testPartSections) === $this->getAssessmentSection()
+            && $this->getTestPart()->getBranchRules()->count() > 0
+        ) {
+            return $this->getTestPart()->getBranchRules();
+        }
+
+        return new BranchRuleCollection();
+    }
+
+    /**
+     * Selects branching rules from the section/subsection.
+     * Branching rules will be selected only if the item or subsection is the last one in the parent section.
+     *
+     * @return ?BranchRuleCollection Returns the branching rules for the last section or null if the element/subsection
+     *                               is not the last.
+     */
+    public function getEffectiveSectionBranchRules(): ?BranchRuleCollection
+    {
+        $sections = $this->getAssessmentSections()->getArrayCopy();
+        $currentSection = array_pop($sections);
+        $currentSectionItems = $currentSection->getSectionParts()->getArrayCopy();
+
+        if (end($currentSectionItems) !== $this->getAssessmentItemRef()) {
+            return null;
+        }
+
+        if ($currentSection->getBranchRules()->count() > 0) {
+            return $currentSection->getBranchRules();
+        }
+
+        $lastSection = $currentSection;
+
+        foreach (array_reverse($sections) as $section) {
```

Comment:
Why do you need to evaluate all the sections but not the current section?

Reply:
Section - has branching rules
- Subsection 1
  - Item 1
- Subsection 2
  - Item 2 - We are here
- Subsection 3
  - Item 3

Flow:
1. Checking `Item 2` for branching rules
   * No branching rules
3. Checking if `Item 2` is the last one in the `Subsection 2`
   * True
5. Checking `Subsection 2` for branching rules
   * No branching rules
7. Checking if `Subsection 2` is the last one in the `Section`
   * False
9. Moving to the `Item 3`
10. Checking `Item 3` for branching rules
    * No branching rules
12. Checking if `Item 3` is the last one in the `Subsection 3`
    * True
14. Checking `Subsection 3` for branching rules
    * No branching rules
16. Checking if `Subsection 3` is the last one in the `Section`
    * True
18. Checking `Section` for branching rules
    * Branching rules exists
20. Return `Section` branching rules

This iteration helps me to go up from the low level subsection to the top level section